src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubDirectDStream.java

```diff
@@ -0,0 +1,268 @@
+
+  private SubscriptionAdminClient subscriptionAdminClient;
+  private final Credentials credentials;
+  private PubSubSubscriberConfig config;
+  private long readDuration;
+  private io.cdap.cdap.etl.api.streaming.StreamingContext context;
+  private final boolean autoAcknowledge;
+  private final transient SerializableFunction<PubSubMessage, T> mappingFn;
+  private StreamingContext streamingContext;
+  private String currentSnapshotName;
+  private boolean takeSnapshot;
+  private String existingSnapshot;
+
+  public PubSubDirectDStream(io.cdap.cdap.etl.api.streaming.StreamingContext context, PubSubSubscriberConfig config,
+                             long readDuration, boolean autoAcknowledge,
+                             SerializableFunction<PubSubMessage, T> mappingFn) {
+    super(context.getSparkStreamingContext().ssc(), scala.reflect.ClassTag$.MODULE$.apply(PubSubMessage.class));
+    this.streamingContext = context.getSparkStreamingContext().ssc();
+    this.config = config;
+    this.readDuration = readDuration;
+    this.context = context;
+    this.autoAcknowledge = autoAcknowledge;
+    this.mappingFn = mappingFn;
+    this.credentials = createCredentials(config.getServiceAccount(), config.isServiceAccountFilePath());
+  }
+
+  @Override
+  public Option<RDD<T>> compute(Time validTime) {
+    LOG.debug("Computing RDD for time {}.", validTime);
+    PubSubRDD pubSubRDD = new PubSubRDD(streamingContext.sparkContext(), readDuration, config, autoAcknowledge,
+                                        credentials);
+    final SerializableFunction<PubSubMessage, T> localMappingFn = mappingFn;
+    // SerializableFunction1<PubSubMessage, T> pubSubMessageTFunction1 = ;
+    RDD<T> mapped = pubSubRDD.map(localMappingFn, scala.reflect.ClassTag$.MODULE$.apply(PubSubMessage.class));
+    return Option.apply(mapped);
+  }
+
+  @Override
+  public void start() {
+    if (config.getTopic() != null) {
+      createSubscriptionIfNotPresent();
+    }
+
+    try {
+      subscriptionAdminClient = buildSubscriptionAdminClient(credentials);
+    } catch (IOException e) {
+      throw new RuntimeException("SubscriptionAdminClient creation failed.", e);
+    }
+
+    // If state with snapshot is present, seek to it.
+    this.existingSnapshot = fetchSnapShot(config.getSubscription(), context);
+    if (this.existingSnapshot != null) {
+      seekSnapshot(config.getProject(), config.getSubscription(), existingSnapshot);
+      // Can use the same snapshot till end of the batch.
+      takeSnapshot = false;
+    } else {
+      takeSnapshot = true;
+    }
+  }
+
+  @Override
+  public void stop() {
+    if (subscriptionAdminClient == null) {
+      return;
+    }
+
+    subscriptionAdminClient.close();
+  }
+
+  @Override
+  public void onBatchStarted(io.cdap.cdap.etl.api.streaming.StreamingContext streamingContext) {
+    LOG.debug("Starting a batch.");
+    if (!takeSnapshot) {
+      this.currentSnapshotName = existingSnapshot;
```

Comment:
do we need existingSnapshot? I may be missing something, but it seems like we can just have the currentSnapshot variable?

Reply:
Removed.